AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -41,68 +42,106 @@ @interface ASVideoNode ()
 
 @implementation ASVideoNode
 
-- (instancetype)init
+//TODO: Have a bash at supplying a preview image node for use with HLS videos as we can't have a priview with those
+
+
+#pragma mark - Construction and Layout
+
+- (instancetype)initWithURL:(NSURL*)url
 {
+  ASDisplayNodeAssertNotNil(url, @"URL must be supplied in initWithURL:");
   if (!(self = [super init])) {
     return nil;
   }
-    
+  
+  _url = url;
+  return [self commonInit];
+}
+
+- (instancetype)initWithAsset:(AVAsset*)asset
+{
+  ASDisplayNodeAssertNotNil(asset, @"Asset must be supplied in initWithAsset:");
+  if (!(self = [super init])) {
+    return nil;
+  }
+  _asset = asset;
+  return [self commonInit];
+}
+
+- (instancetype)commonInit
+{
   _previewQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0);
   
   self.playButton = [[ASDefaultPlayButton alloc] init];
-  
   self.gravity = AVLayerVideoGravityResizeAspect;
-  
   [self addTarget:self action:@selector(tapped) forControlEvents:ASControlNodeEventTouchUpInside];
-    
+  
   return self;
 }
 
-- (void)interfaceStateDidChange:(ASInterfaceState)newState fromState:(ASInterfaceState)oldState
+- (instancetype)init
 {
-  if (!(newState & ASInterfaceStateVisible)) {
-    if (oldState & ASInterfaceStateVisible) {
-      if (_shouldBePlaying) {
-        [self pause];
-        _shouldBePlaying = YES;
-      }
-      [(UIActivityIndicatorView *)_spinner.view stopAnimating];
-      [_spinner removeFromSupernode];
-    }
-  } else {
-    if (_shouldBePlaying) {
-      [self play];
-    }
-  }
+  ASDisplayNodeAssertNotSupported();
+  return nil;
 }
 
-- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+- (instancetype)initWithViewBlock:(ASDisplayNodeViewBlock)viewBlock didLoadBlock:(ASDisplayNodeDidLoadBlock)didLoadBlock
 {
-  if ([[change objectForKey:@"new"] integerValue] == AVPlayerItemStatusReadyToPlay) {
-    if ([self.subnodes containsObject:_spinner]) {
-      [_spinner removeFromSupernode];
-      _spinner = nil;
+  ASDisplayNodeAssertNotSupported();
+  return nil;
+}
+
+- (ASDisplayNode*) constructPlayerNode {
+  ASDisplayNode* playerNode = [[ASDisplayNode alloc] initWithLayerBlock:^CALayer *{
+    AVPlayerLayer *playerLayer = [[AVPlayerLayer alloc] init];
+    if (!_player) {
+      [self constructCurrentPlayerItemFromInitData];
+      _player = [AVPlayer playerWithPlayerItem:_currentPlayerItem];
+      _player.muted = _muted;
     }
-  }
+    playerLayer.player = _player;
+    playerLayer.videoGravity = [self gravity];
+    return playerLayer;
+  }];
   
-  if ([[change objectForKey:@"new"] integerValue] == AVPlayerItemStatusFailed) {
-    
+  return playerNode;
+}
+
+- (void) constructCurrentPlayerItemFromInitData {
```

Comment:
Minor, but please ensure all of your methods have no space after the ), and have the { on the next line (see others in the file / other files for reference)